ユーザモデルを追加

a5ob7r opened the pull request:
# 概要

APIにおけるユーザに対応するモデルと対応するテストを作成する。コントローラについては別PRにて実装する予定である。

a5ob7r committed:
Add bcrypt

a5ob7r committed:
Enable pgcrypto extension for UUID

a5ob7r committed:
Add User model

a5ob7r committed:
Add tests for User model

a5ob7r commented:
#26 のマージで Gemfile.lock が更新されたからこのブランチでコンフリクトするだろうなと思ったけどそんなことはなかった。

tim-toga commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,9 @@
+  factory :user do
+    name { 'MyString' }
+    email { 'MyString' }
+    password_digest { 'MyString' }
```
passwordもdefaultで定義されてた方がいいのでは？意図的にはずしてる？

tim-toga commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,181 @@
+                        password_confirmation: 'a_little_bit_long_password'
+  end
+
+  it { is_expected.to be_valid }
```
これから入るチームでは基本的にcontextは階層で作るようにしてます
```ruby
context 名前がnil
end
context 名前がstring
  context 空文字
  end
  context 空文字でない
    context 一文字
    end
    ...
  end
end
```

例えば、名前に使用不可な文字を追加したいとか、メールアドレスのブラックリストを追加したいとか条件が増えた場合に、今のspecの構造だと全てのcontextが並列なのでどこに追加してもいいようにも見えます（雰囲気で名前とかメールで固まってるのはわかるけど）

KT-tim replied:
ここは今は必須じゃないからチーム入る時に覚えてればいいかなって感じがします
例えば’特定の権限を持ったユーザーだけ超短い名前を許可する’みたいな処理があるならcontextをネストして書く必要があるんですが、今回の場合は将来的な追加のしやすさを考えているだけなのでまだ好みの差と言えると思います

tim-toga replied:
@KT-tim 
はい。修正の提案じゃないです。どちらかというと`チームに入る時に覚える`というよりは、`いい機会だからチームのやり方を紹介してる`というニュアンスです。

tim-toga commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,181 @@
+
+  it { is_expected.to be_valid }
+
+  context '名前がnilのとき' do
```
名前、メール、パスワードごとにdescribeで囲う方が見やすいな

a5ob7r replied:
こちらのコメントも考慮すると、属性 -> 値の型 -> 値の順の階層を作るといったところなのでしょうか。

https://github.com/sls-training/2023-API-training-server/pull/25#discussion_r1222459218

tim-toga replied:
ですね

tim-toga commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,181 @@
+    it { is_expected.to be_valid }
+  end
+
+  context '名前の文字数が最長のとき' do
```
微妙なところだけど、`64文字のとき`と`65文字以上のとき`の方がわかりやすいかも

a5ob7r replied:
元々はそのように書いていたのですが、それだとテスト実行時の出力を眺めた時に、64とか65とかいう数字がどういった意味合いなのかわからないのでは、と思って、わざと値の意味合いで説明を書きました。まあ、この2つが並んでいれば、そこが値の境界なのだろうと想像はできるとは思っています。

tim-toga commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,181 @@
+      user.name = 'a' * 65
+    end
+
+    it { is_expected.not_to be_valid }
```
`be_invalid`っていうのもあるよ

tk0miya replied:
雑談: 個人的には not_to を使わずに済むときは not_to を使いたくないので be_invalid の方が好みかなー

a5ob7r replied:
おお、知りませんでした。使ってみます。

a5ob7r replied:
これくらいなら、指摘してくれそうなCopがありそうだけどないのかな。

tim-toga commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,181 @@
+    it { is_expected.not_to be_valid }
+  end
+
+  context '名前がスペースのみで構成されているとき' do
```
空文字の検証と被ってない？

a5ob7r replied:
空文字が文字列長が0の文字列であるのに対して、ご指摘の箇所は、文字列長が1以上でかつ含まれる文字がスペースのみの場合のテストという認識です。なので被ってはいないと思います。まあ、内部的には #blank を使って検証するのでどちらの場合も同じ設定で検出可能ではありますが、テストではそういった内部実装は特に考慮しないものだと思っています。

epaew replied:
境界値テストをちゃんと理解してるマンだ、いいね 👍

まあ、厳密にやりすぎると「自分の書いたコードのテスト」じゃなくて、「Rails のテスト」になりかねないので、無駄も大きくなっちゃうんだけどね。

a5ob7r replied:
今気がつきましたが、スペースを1つだけ代入するなら積演算は不要ですね。削除します。

```ruby
      user.name = ' ' * 1
```

play-toga replied:
blank?が意図通りに機能してるか？のテストに見えますね。nil.blank? と ''.blank? と '   '.blank? をそれぞれテストする意味ってありますかね？

epaew replied:
request spec / worker spec はブラックボックステストで書きたいけど、  
その他の単体テストはホワイトボックステストでもいいんじゃないかなー、が僕の感覚です

play-toga replied:
たしかに、modelのspecはホワイトボックスでいいかもしれないですね。
あってもいいけど、無くていい派になりました

a5ob7r replied:
> ホワイトボックステスト

ふむふむ。

https://ja.wikipedia.org/wiki/ホワイトボックステスト

tim-toga commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,181 @@
+
+RSpec.describe User, type: :model do
+  subject(:user) do
+    described_class.new name: 'foobar', email: 'foobar@foo.bar', password: 'a_little_bit_long_password',
```
name, emailのそれぞれをlet変数で定義する方が見やすいですね
```ruby
subject(:user) do
  described_class.new
    name: name,
    email: email,
end

context 名前がnilのとき
  let(:name) { nil }

  it {}
end
```

a5ob7r replied:
見やすさについては、慣れの問題も大きそうなので、何とも言いづらいですね。この場合は、手続的か宣言的かの違いなのかな。

ただ、提案してくださった記述の仕方は、 Hash Value Omission と相性が良さそうですね。
https://bugs.ruby-lang.org/issues/14579

a5ob7r replied:
宣言的な書き方のほうが、 `名前がnilのとき` みたいな説明文とも親和性があっていいですね。見やすさについては、やはり判断が分かれるところですが、RSpecのようなドキュメントチックな書き方をする場合は、宣言的な書き方のほうが文言と書き方の統一感があって良さげだと感じました。

tim-toga commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
```
factorybot定義してるけど使ってないような？

a5ob7r replied:
自動生成されたものをそのままにしてました。ご指摘のとおり、本PRでは利用していないので削除します。

tk0miya commented on .rubocop.yml:
```diff
@@ -19,3 +19,9 @@ AllCops:
     - public/**/*
     - storage/**/*
+
+RSpec/ExampleLength:
```
感想: こんな設定あったんだね。しらなかった。便利！

a5ob7r replied:
こちらから拝借しました。
https://github.com/sls-training/2023-rails-sample/blob/6c948131445ab412ef81ba0f2ef400efe0ac0b72/.rubocop_todo.yml#L8

a5ob7r committed:
Remove an unused factory

a5ob7r committed:
Remove an unnecessary product operator

tk0miya reviewed (commented):
バリデーション周りのテストって難しいですよねえ。

僕らが書いたコードは validates 文の定義だけなので、

* validates 宣言がちゃんと書けているかは確認したい
* どうバリデーションされるかは Rails 次第なので任せたい

というのが本音ですね。
途中でコメントされている空文字と空白文字だけの違いというのは `presence: true` の中身のテストなので、Rails 本体のテストであると言ってもいいんですよね。

ここらへん、テストコードを書いても得るものが少ないので、あまり細かくテストを書かない (まるごと端折ってしまうことすらある) という感じなので、落とし所が見えなくて厳しいです。

※ まるごとテストを書かない場合、宣言が間違っているのに気づけないのですが、それは人間がテストをして確認するということで担保していますね。

a5ob7r committed:
Change `not_to be_valid` to `to be_invalid`

a5ob7r committed:
Change to definitive expressions

a5ob7r committed:
Change a simple fuzzy testing to just one

If need to do fuzzy testing, this is too few times. This should be just
a black-box test for email formats.

a5ob7r committed:
Use negative matchers instead of not_to

a5ob7r committed:
Add a describe for each of properties

KT-tim commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,158 @@
+    end
+  end
+
+  describe '.password_confirmation' do
```
そういえばpasswordとpassword_confirmationは一致してなくてもいいものですか？

a5ob7r replied:
以前、 password_confirmation が nil だとパスワード関連のバリデーションが実行されない的なことを言いましたが、改めて確認してみると、どうやら勘違いのようでした。ちゃんとバリデーションされてますね。というわけで、 password_confirmation 属性に関するテスト自体が不要になりました。よって削除します。

a5ob7r replied:
それにしても前はどうして勘違いしてしまったのだろう。妖怪何もしていないのに勝手に直ったに遭遇したようです。

a5ob7r committed:
Remove an unnecessary validation for .password_confirmation

It was my misunderstanding, it's embarrassing...

a5ob7r committed:
Remove redundant validations for the password attribute

a5ob7r committed:
Fix some incorrect tests

a5ob7r committed:
Remove a redundant test

It's enough to check the validation.

a5ob7r committed:
Fix an explanation for some tests

a5ob7r commented:
一通り修正しました。確認をお願いします。

tim-toga commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,145 @@
+      it { is_expected.to be_valid }
+    end
+
+    it '不正なフォーマットを拒否する' do
```
ここだけcontextの階層じゃないの違和感ある

a5ob7r replied:
統一感がないのはその通りなのですが、 context を使おうとすると、逆に違和感があるのではと考えて採用しませんでした。 context 内部の it の適切な説明文が思いつかなかったということがあります。↓の　検証に失敗する といった、自明な説明でもいいんですかね。

```ruby
context '不正なフォーマットのとき'
  it '検証に失敗する' do
    ...
    ...
  do
end
```

tim-toga replied:
はい。自明でもcontextやitの階層が揃ってるほうがみやすいので、そうすることが多いです（もしかしたら"不正"を具体的に書くと自明感が下がるかも）

a5ob7r replied:
自明と感じていたのは、 context の説明ではなく it の説明のほうでした。これは、もともと英語で書いていたときに、同様の説明を記述する際に、単に fails と記述していたことが理由でした。ただ、今回改めて日本語で書いてみると、そこまで違和感がないかも、と感じてきました。

a5ob7r replied:
せっかくなので修正してみます。

tk0miya replied:
いいと思う。

rspec は BDD (Behavior Driven Development) というテスト手法に影響を受けているツールなんだけども、
BDD ではテストを以下のように整理します。

* Given: どういう状態のときに
* When: 何をしたら
* Then: どうなった

rspec ではそれぞれ以下のように対応します。

* Given: context とそれを具現化する before や let など
* When: subject のこと
* Then: it のこと

つまり、context には「どういう状態のときに」を書くべきで、「どうなった」を書くのは不適当です。
そのため、context には「〜〜のとき」「〜〜の場合」「〜〜をしたら」といった状況を表す言葉が入るのが適当で、it には「〜〜になる」「〜〜すること」といった結論を表す言葉が入ります。

上でコメントをしたけど `rspec --format=document` するとこの様子がわかりやすくなるはずです。

tim-toga commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,145 @@
+      it { is_expected.to be_invalid }
+    end
+
+    context '大文字小文字違いのメールアドレスを持つユーザが存在する場合' do
```
どちらかというと「大文字で送信したメールアドレスが小文字で保存されていること」かな

a5ob7r replied:
うーん、それだと保存時の処理をテストしていることになりませんかね。それだとバリデーションの結果ではなく、保存後のオブジェクトのメールアドレスのケースを比較する場合の説明のような気がします。

a5ob7r replied:
というか、 context というより it 向きの説明ですかね。

tim-toga replied:
あ、たしかにitの説明文書いてた。。

sample@foo.com が存在する場合に SAMPLE@foo.com をバリデーションするのが目的ではなく、SAMPLE@foo.comをsample@foo.comで保存するのが目的かと思ったので。なので、it節も`be_invalid`ではなく `eq samlple@foo.com` とかになるかなー。
（+1してるのでsatoくんのタイミングでマージしてください）

a5ob7r replied:
今回は、一意な属性であるメールアドレスが case-insensitive であることをテストできればいいので、保存時の形式については不問としています。

tk0miya commented on app/models/user.rb:
```diff
@@ -0,0 +1,11 @@
+  before_save { email.downcase! }
+
+  has_secure_password validations: false
+  validates :password, presence: true, length: { minimum: 16, maximum: 128 }
```
僕の経験では、フィールドごとに定義を寄せるのではなく、メソッドの種類ごとに定義を寄せる方が見やすいと思います。
validates は validates で集める、イベントフック(before_*, after_* など) も一箇所に集める…という感じです。

a5ob7r replied:
そういえば、Railsチュートリアルでもそのように実装されていましたね。

個人的には、フィールド単位というか、論理的に関連するものをまとめておくほうが好みです。ですが、メソッドの種類でまとめるほうが、まとめかたの基準も明確だし、メソッド同士の順序もそれぞれ自由に変更できるので、開発を継続していく上ではそちらのほうが理にかなっているなと感じました。単純に同じような文言が並ぶので、整頓されていてきれいに見える、というのもありますね。

ただ、今回は今のところ、ここから変更する明確な必要性はないので、そのままにしておこうと思います。ただ、↑のようにメソッド単位でまとめるほうが理にかなっているということは頭に入れておきます。

tk0miya commented on app/models/user.rb:
```diff
@@ -0,0 +1,11 @@
+  validates :name, presence: true, length: { minimum: 1, maximum: 64 }
+
+  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }, uniqueness: true
+  before_save { email.downcase! }
```
雑談: アプリの仕様次第だけど、小文字に統一するのであれば、ユーザーを検索するときも小文字にしたクエリが必要ですね。ログインするときとか、トークンを払い出すときとか、 User を検索するときには email を小文字に変換する必要がありますね。

「小文字にして検索する」という処理をどこに書くのかは議論が分かれて、設計が面白そうなポイントです。
今回は API 定義で結構時間を使ったので、ここについては深掘りしないでおきますが、思考実験としてはとても面白そうなトピックだと思いました。

tk0miya commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,145 @@
+  let(:email) { 'foobar@foo.bar' }
+  let(:password) { 'a_little_bit_long_password' }
+
+  it { is_expected.to be_valid }
```
ここ、context がなにもないのは変だと思うなあ。
すべてのカラムに適切な値が入っているとき、みたいな context ですよね、これの意図は。

rspec には `--format documentation` というオプションがあるので、それをつけて出力結果を見てみると面白いと思います。

a5ob7r replied:
なるほど、あってもなくても結果に違いはないので、ここではcontextは使っていませんでしたが、見やすさを考えると、ある種の空のcontextがあってもいいかもしれないですね。

tk0miya commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,145 @@
+
+    context '大文字小文字違いのメールアドレスを持つユーザが存在する場合' do
+      before do
+        another_user = user.dup
+        another_user.email = user.email.upcase
+        another_user.save!
```
ここは dup を使わず、新しいインスタンスを作るのがいいと思う。
他の理由で保存できなかった、というのは本末転倒だよね。

a5ob7r replied:
なるほど、それもそうですね。変更します。

a5ob7r commented:
Approve数が2つになったのでマージします。ありがとうございました。
